### ISSUE TITLE
feat(got): port GOT-OCR 2.0 (SAM ViT-B tower reuse + linear projector + Qwen2-0.5B decoder, QWen tiktoken specials)

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1359]
## Summary

GOT-OCR 2.0 (`stepfun-ai/GOT-OCR2_0`, `model_type: "GOT"`, `architectures: ["GOTQwenForCausalLM"]`; MLX conversions `mlx-community/GOT-OCR2_0-{bf16,8bit,4bit}`) is a 0.58B OCR model: a SAM-style ViT-B tower (1024x1024 input, window attention 14, global blocks 2/5/8/11, decomposed relative position bias, two-conv neck, two stride-2 compressor convs) that yields 256 tokens of width 1024, a single `Linear(1024, 1024)` projector, and a Qwen2-0.5B decoder with attention bias and tied embeddings. mlxcel already ships the identical tower for DeepSeek-OCR (`SamEncoder`), the Qwen2 decoder, and a tiktoken loader; what is missing is detection, the key layouts, the `<img>` / `<imgpad>` / `</img>` prompt framing with its fixed system prompt, the stop-token set, and the QWen tiktoken special-token table (the current tiktoken loader hardcodes the HunYuan table, which gives `<|im_end|>` the wrong id).

## Current behavior

- `src/models/detection.rs` lowercases `model_type` (line 147) and has no `"got"` arm.
- `src/vision/encoders/deepseekocr_sam.rs:34-60, 356-493` defines `SamConfig { embed_dim 768, num_heads 12, depth 12, window_size 14, global_attn_indexes [2,5,8,11], out_chans 256, final_out_chans 1024, grid 64 }` and `SamEncoder::from_weights(weights, prefix, config)` reading `{prefix}.patch_embed.proj.{weight,bias}`, `{prefix}.pos_embed`, `{prefix}.blocks.N.{norm1,norm2,attn.qkv,attn.proj,attn.rel_pos_h,attn.rel_pos_w,mlp.lin1,mlp.lin2}`, `{prefix}.neck.{0,1,2,3}`, `{prefix}.net_2.weight`, `{prefix}.net_3.weight`, with the shape-gated `conv_channels_last` transpose (line 373). `forward` takes `[B, 1024, 1024, 3]` and returns `[B, 16, 16, 1024]`. This is the GOT tower exactly; only the key prefix and the neck key names differ between the two checkpoint layouts.
- `src/tokenizer/tiktoken.rs:52-130` (`TiktokenTokenizer::from_file`) builds the special-token table from `build_special_token_list()` = `<|endoftext|>, <|startoftext|>, <|bos|>, <|eos|>, <|pad|>, <|extra_0..204|>` starting at `encoder.len()`, then overrides from `tokenizer_config.json` `added_tokens_decoder`. GOT ships `qwen.tiktoken` (151643 ranks) with an empty `added_tokens_decoder`, so the loader would assign `<|startoftext|>` = 151644 and `<|bos|>` = 151645 where GOT needs `<|im_start|>` = 151644 and `<|im_end|>` = 151645, and it has no entries at all for `<ref>` 151851, `</ref>` 151852, `<box>` 151853, `</box>` 151854, `<quad>` 151855, `</quad>` 151856, `<img>` 151857, `</img>` 151858, `<imgpad>` 151859. The pre-tokenization regex (`HUNYUAN_PAT`, line 29) is identical to the QWen one, so only the specials need a second table.
- `src/models/qwen2.rs` re-exports `Llama3Model` as `Qwen2Model` with `attention_bias` and `tie_word_embeddings` honored (`src/models/llama3.rs:49, 64, 1108`).

## Expected behavior

```
./target/release/mlxcel generate -m models/GOT-OCR2_0-4bit --image page.png -p "OCR:" -n 1024
./target/release/mlxcel generate -m models/GOT-OCR2_0-4bit --image page.png -p "format:" -n 2048
```

resize the image to 1024x1024 (bicubic, no aspect preservation), normalize with CLIP mean/std, run the tower to `[1, 256, 1024]`, project, scatter the 256 features onto the `<imgpad>` positions of the fixed conversation, and decode until `<|im_end|>` (151645) or `<|endoftext|>` (151643). `OCR:` returns plain text; `format:` returns markdown/LaTeX-formatted text; `OCR with format: {"box": [y1, x1, y2, x2]}` and `OCR with format: {"color": "red"}` are passed through as instructions (fine-grained modes; coordinates are normalized to 0..1000).

### config.json (flat; the text config is the root)

`model_type "GOT"`, `hidden_size 1024`, `num_hidden_layers 24`, `num_attention_heads 16`, `num_key_value_heads 16`, `intermediate_size 2816`, `rms_norm_eps 1e-6`, `rope_theta 1000000.0`, `max_position_embeddings 32768`, `vocab_size 151860`, `tie_word_embeddings true`, `use_sliding_window false`, `bos_token_id 151643`, `eos_token_id 151643`, `image_token_len 256`, `im_start_token 151857` (`<img>`), `im_end_token 151858` (`</img>`), `im_patch_token 151859` (`<imgpad>`), `use_im_start_end true`. No `vision_config` (tower geometry is fixed: `img_size 1024`, `patch_size 16`, `embed_dim 768`, `depth 12`, `num_heads 12`, `out_chans 256`, `out_dim 1024`, `window_size 14`, `global_attn_indexes (2,5,8,11)`), no `text_config`, no `quantization` on the original; converted checkpoints add `quantization {group_size 64, bits 4}`.

### Prompt

The checkpoint ships no chat template. The conversation is fixed:

```
<|im_start|>system
        You should follow the instructions carefully and explain your answers in detail.<|im_end|><|im_start|>user
<img><imgpad>x256</img>
{instruction}<|im_end|><|im_start|>assistant
```

(the system line starts with eight literal spaces after the newline; `<|im_end|>` has no trailing newline). If the instruction already contains `<|im_end|>` or `<|im_start|>user\n` it is used verbatim; if it contains `<image>` that placeholder is replaced by the image block, otherwise the block is prepended. No BOS is added. Stop ids: `[151643, 151645]` (the config declares only 151643; without 151645 generation never stops).

### Image processor

Resize to exactly 1024x1024 with bicubic filtering, `pixel / 255`, normalize `(x - mean) / std` with `mean [0.48145466, 0.4578275, 0.40821073]`, `std [0.26862954, 0.26130258, 0.27577711]`; emit `[1, 3, 1024, 1024]` (the tower takes NHWC, transpose once before the call).

### Weight keys

Original layout (`stepfun-ai/GOT-OCR2_0`):

```
model.vision_tower_high.patch_embed.proj.{weight,bias}   # weight [768, 3, 16, 16]
model.vision_tower_high.pos_embed                        # [1, 64, 64, 768]
model.vision_tower_high.blocks.{0..11}.{norm1,norm2}.{weight,bias}
model.vision_tower_high.blocks.N.attn.{qkv,proj}.{weight,bias}
model.vision_tower_high.blocks.N.attn.{rel_pos_h,rel_pos_w}  # [27, 64] windowed, [127, 64] global
model.vision_tower_high.blocks.N.mlp.{lin1,lin2}.{weight,bias}
model.vision_tower_high.neck.0.weight   # [256, 768, 1, 1]
model.vision_tower_high.neck.1.{weight,bias}
model.vision_tower_high.neck.2.weight   # [256, 256, 3, 3]
model.vision_tower_high.neck.3.{weight,bias}
model.vision_tower_high.net_2.weight    # [512, 256, 3, 3]
model.vision_tower_high.net_3.weight    # [1024, 512, 3, 3]
model.mm_projector_vary.{weight,bias}   # Linear 1024 -> 1024
model.embed_tokens.weight, model.layers.N.*, model.norm.weight   # Qwen2, q/k/v with bias
lm_head.weight                          # tied copy, ignore
```

Converted layout (`mlx-community/GOT-OCR2_0-*`): `vision_tower.{patch_embed.proj, pos_embed, blocks.N.*, conv1, norm1, conv2, norm2, net_2, net_3}` (neck.0/1/2/3 renamed to conv1/norm1/conv2/norm2), `multi_modal_projector.{weight,bias}`, `language_model.model.*` (+ `.scales/.biases` on quantized linears and on `embed_tokens`), no `lm_head`. Conv weights are already `[O, kH, kW, I]`; the tower is not quantized.

## Implementation plan

1. **Detection.** `src/models/detection.rs`: `"got" => Ok(ModelType::GotOcrVLM)` (the raw value is `GOT`; the existing lowercasing at line 147 makes this arm fire). `src/models/mod.rs`: enum, supported list, description `("GOT-OCR 2.0 (SAM ViT-B + linear projector + Qwen2-0.5B)", "Other VLM")`. Test in `src/models/detection_tests.rs` with `{"model_type": "GOT", "hidden_size": 1024}`.

2. **QWen tiktoken specials (`src/tokenizer/tiktoken.rs`).** Add `fn build_qwen_special_token_list() -> Vec<String>` = `<|endoftext|>, <|im_start|>, <|im_end|>, <|extra_0|>..<|extra_204|>, <ref>, </ref>, <box>, </box>, <quad>, </quad>, <img>, </img>, <imgpad>` (ids `encoder.len() + i`, so 151643..151859 for the 151643-rank vocab), selected when `tokenizer_config.json` has `tokenizer_class == "QWenTokenizer"` (keep the HunYuan table otherwise). Keep the `added_tokens_decoder` override after either table. Tests in the existing `mod tests` of `src/tokenizer/tiktoken.rs` (line 306): `qwen_tokenizer_class_selects_qwen_specials` (a 3-rank synthetic vocab yields `<|im_end|>` = 5 and `<imgpad>` = 219), `hunyuan_table_unchanged_without_qwen_class`.

3. **Config (`src/vision/got_ocr.rs`, new).** `GotOcrConfig` deserialized from the flat root: the Qwen2 fields go through `models::llama3::ModelArgs` (set `model_type` to `"qwen2"` before parsing so the Llama-family loader takes the attention-bias branch, and inject the top-level `quantization`), plus `image_token_len` (256), `im_start_token` (151857), `im_end_token` (151858), `im_patch_token` (151859), `eos_token_id`. Constant `GotVisionConfig` = `SamConfig::default()` (`final_out_chans 1024`, `grid 64`).

4. **Key canonicalization (`src/loading/vlm_got_ocr.rs`, new).** `fn canonicalize_got_keys(WeightMap) -> WeightMap`: `model.vision_tower_high.neck.0.` -> `vision_tower.conv1.`, `.neck.1.` -> `.norm1.`, `.neck.2.` -> `.conv2.`, `.neck.3.` -> `.norm2.`, other `model.vision_tower_high.X` -> `vision_tower.X`, `model.mm_projector_vary.X` -> `multi_modal_projector.X`, `model.X` -> `language_model.model.X`, `lm_head.X` -> `language_model.lm_head.X`; pass converted keys through unchanged; drop `language_model.lm_head.weight` when `tie_word_embeddings`. Since `SamEncoder::from_weights` reads `neck.0..3`, either give `SamEncoder` a `NeckKeys { conv1, norm1, conv2, norm2 }` parameter (default `neck.0..3`, GOT passes `conv1/norm1/conv2/norm2`) or map the converted names back to `neck.N` in the canonicalizer; the second is smaller. Conv layout is handled by the existing `conv_channels_last` gate for every conv (`patch_embed.proj.weight`, `neck.0`, `neck.2`, `net_2`, `net_3`; all four cases pass the gate correctly: `[256,768,1,1]` and `[256,256,3,3]` transpose, `[256,1,1,768]` and `[256,3,3,256]` stay). Unit test `canonicalize_got_keys_maps_both_layouts_to_one` with the 2-layout fixtures above, and `sanitize_is_idempotent` (running twice equals once).

5. **Runtime (`src/vision/got_ocr.rs`).** `GotOcrVlModel { text_model: Qwen2Model, vision_tower: SamEncoder, projector: UnifiedLinear /* multi_modal_projector, bias */, image_token_len: 256, im_start_token_id, im_end_token_id, im_patch_token_id, eos_token_ids: vec![151643, 151645] }`. `get_input_embeddings(input_ids, pixel_values)`: NCHW -> NHWC, cast to the tower dtype, `vision_tower.forward` -> `[1, 16, 16, 1024]` -> reshape `[1, 256, 1024]` -> `projector.forward` -> `merge::merge_llava(im_patch_token_id, ...)` (the count of `<imgpad>` positions must equal 256 per image). `LanguageModel` by delegation to `Qwen2Model` (batching and paged decode pass through; the tower runs once per request in `prepare_vlm_inputs`). Processor: `GotOcrImageProcessor { image_size 1024, mean, std }` in `src/vision/processors/got_ocr.rs` using `FilterType::CatmullRom` and `resize_exact`.

6. **Prompt builder (`src/multimodal/got_ocr_prompt.rs`, new).** `build_got_prompt(instruction, has_image) -> String` implementing the fixed conversation and the `<image>` / prepend rule; `insert` the 256-token block as text (the tokenizer resolves `<img>`, `<imgpad>`, `</img>` to the special ids once step 2 lands), so the runtime arm only needs to tokenize the built string. Tests: `plain_instruction_is_wrapped_with_system_and_image_block`, `pre_templated_instruction_passes_through`, `image_placeholder_is_replaced_in_place`, `block_has_256_imgpad_tokens`.

7. **Dispatch.** `LoadedModel::GotOcrVLM` (`src/loaded_model.rs`), `VlmRuntimeRef::GotOcr` (`src/loaded_model_capabilities.rs`), a `vlm_runtime.rs` arm that builds the prompt, tokenizes (no BOS), preprocesses, calls `get_input_embeddings`, and reports `VlmPreparationSummary::GotOcr { image_tokens: 256 }` (print in `src/commands/generate_vlm.rs`). The server path: `/v1/chat/completions` image parts reach the same arm; the user text is the instruction, and the fixed system prompt replaces any client system message (document this).

8. **Output suppression.** `[151857, 151858, 151859]`.

9. **Docs.** `docs/supported-models.md`: add GOT-OCR 2.0 with the four prompting modes.

## Validation

(a) Unit tests from steps 1, 2, 4, 6 plus `src/vision/got_ocr_tests.rs::tower_output_is_256_tokens_of_1024` on random weights with `SamConfig::default()`.

(b) Real checkpoint: `mlx-community/GOT-OCR2_0-4bit` (0.45 GB) and `mlx-community/GOT-OCR2_0-bf16` (about 1.4 GB); the original `stepfun-ai/GOT-OCR2_0` for the hub key layout.

```
magick -size 1024x512 xc:white -pointsize 48 -fill black -annotate +80+200 "GOT OCR two point zero" /tmp/got_page.png
./target/release/mlxcel generate -m models/GOT-OCR2_0-bf16 --image /tmp/got_page.png -p "OCR:" -n 64
./target/release/mlxcel generate -m models/GOT-OCR2_0-4bit --image /tmp/got_page.png -p "format:" -n 128
./target/release/mlxcel generate -m models/GOT-OCR2_0-bf16 --image tests/fixtures/test_image.png -p "OCR:" -n 64
```

Acceptance: finite logits; the rendered sentence is transcribed verbatim under `OCR:`; generation stops on `<|im_end|>` without emitting it; the preparation summary reports 256 image tokens; both key layouts load (run the bf16 conversion and the original `stepfun-ai` checkpoint); greedy token-exact match for the first 64 tokens against the checkpoint's own `modeling_GOT.py` on CPU for the rendered page.

## Acceptance criteria

- [x] `model_type: "GOT"` resolves to `ModelType::GotOcrVLM`
- [x] Tiktoken loader selects the QWen special table for `tokenizer_class == "QWenTokenizer"` (`<|im_end|>` = 151645, `<imgpad>` = 151859) and leaves HunYuan unchanged
- [x] Original and converted key layouts both load through one canonicalizer; sanitize is idempotent
- [x] Fixed conversation template, 256 `<imgpad>` block, and stop set `[151643, 151645]` implemented with tests
- [x] `OCR:` transcribes a rendered page on the real checkpoint; 64-token greedy parity
- [x] docs/supported-models.md updated
- [x] detection table in src/models/detection.rs updated with a test
- [ ] cargo test --workspace --profile test-fast --features metal,accelerate passes (not runnable on the implementation host, which is Linux aarch64 / CUDA with no Metal and no Accelerate; unrun, left to the merge gate. The CUDA-scoped equivalents were run and are listed in PR #1744.)
- [x] cargo clippy --workspace --all-targets -- -D warnings and cargo fmt --all -- --check pass (run locally as `--features cuda`, which is the only backend this host builds; the feature-free spelling is covered by the `cargo-clippy` CI job on PR #1744)

## Out of scope

- Multi-page / multi-crop ("multi-crop OCR") and the rendered-HTML output helpers from the model card.
- Multi-image prompts (one image per request; the decoder was trained with a single 256-token block).
